subprojects/core/src/main/java/org/gradle/caching/internal/tasks/TarTaskOutputPacker.java

```diff
@@ -312,15 +328,23 @@ private void unpackPropertyEntry(ResolvedTaskOutputFilePropertySpec propertySpec
             outputFile = new File(propertyRoot, childPath);
         }
 
+        String internedPath = stringInterner.intern(outputFile.getAbsolutePath());
+        RelativePath relativePath = root ? RelativePath.EMPTY_ROOT : RelativePath.parse(!isDirEntry, childPath);
         if (isDirEntry) {
             FileUtils.forceMkdir(outputFile);
+            if (!root) {
+                fileSnapshots.put(propertyName, new DirectoryFileSnapshot(internedPath, relativePath, false));
```

Comment:
Your test does not check if that are no overlapping outputs. Detection only happens if task output caching is enabled.
Your test did test if the task is up to date. This means that all the recorded output snapshots are contained in the output file snapshot taken by up to date checks. In order to check that all the output file snapshots taken by up to date checks are contained in the recorded snapshot from history, you need to check that there have been no overlapping outputs detected. This currently fails, see my comment on the test.